gdc_ng_models/models/study.py

```diff
@@ -0,0 +1,134 @@
+"""Data models to describe studies.
+
+Studies will be used for the single-study controlled-access feature in gdcapi.
+These models provide a way to associate programs and projects to studies.
+"""
+
+from sqlalchemy.ext import declarative
+from sqlalchemy.sql import schema, sqltypes
+
+from gdc_ng_models.models import audit
+
+Base = declarative.declarative_base()
+
+
+class Study(Base, audit.AuditColumnsMixin):
+    """A study for use with single-study controlled-access.
+
+    Attributes:
+        id: A unique identifier for the study.
+        name: A unique, informational name for the study.
+        created_date: The date and time when the record is created.
+        updated_date: The date and time when the record is updated.
+    """
+
+    __tablename__ = "study"
+    id_seq = schema.Sequence(name="study_id_seq", metadata=Base.metadata)
+    id = schema.Column(sqltypes.Integer, nullable=False, server_default=id_seq.next_value())
+    name = schema.Column(sqltypes.Text, nullable=False)
+
+    __table_args__ = (
+        schema.PrimaryKeyConstraint("id", name="study_pk"),
+        schema.Index("study_name_idx", "name", unique=True),
+    )
+
+    def __repr__(self):
+        return "<Study(id={id}, name='{name}', created_date={created_date}, updated_date={updated_date})>".format(
+            id=self.id,
+            name=self.name,
+            created_date=self.created_date.isoformat(),
+            updated_date=self.updated_date.isoformat(),
+        )
+
+    def to_json(self):
+        return {
+            "id": self.id,
+            "name": self.name,
+            "created_date": self.created_date.isoformat(),
+            "updated_date": self.updated_date.isoformat(),
+        }
+
+
+class StudyProgram(Base, audit.AuditColumnsMixin):
```

Comment:
Do we need bidirectional relationships? eg on an instance of study, should we be able to do `study.study_programs` and the reverse `study_program.study`

Reply:
Initially, I don't think so.  We'll be using these table to check if a requested study shape is subset of a defined study.  So, the queries are going to existence checks.  If we end up with a use case where a `relationship` would be useful, we can add it at that point.  

The most important part of this PR is the schema, as that is harder to change.

Reply:
I'm sure we'll need it. instead of doing multiple queries we then can just do one joined query and get the data quicker.

Reply:
I imaged that we'd query `StudyProgram` and `StudyProgramProject` to build all of the `StudyRuleProgram` objects, like:
```
study_programs = db.query(study.StudyProgram)...
# build StudyRulePrograms with whole_project=True from study_programs
study_programs_projects = db.query(study.StudyProgramProject)...
# build StudyRulePrograms with whole_project=false from study_programs_projects
```

Are you suggesting that we add two relationships to `Study`, one for `StudyProgram` and one for `StudyProgramProject`?   Then do something like:
```
studies = db.query(study.Study)...
for study in studies:
    if study.programs:
        # build StudyRulePrograms with whole_program=True from study.programs
    else if study.programs_projects:
        # build  StudyRulePrograms with whole_program=False from study.programs_projects
```

Both approaches work.  I suppose the question is which on is preferred?  I don't have a preference.

Reply:
I just remembered that a study can contain multiple programs and a combination of whole and partial programs.  Having the `relationship`s would make the rule building code simpler as it'd implicitly group by study id.

Reply:
We can query them individually as suggested above, but kind of defeats the need for using an ORM layer which provides these features out of the box with ways of specifying if they are eager or lazy loaded. If there is no use case for bidirectional relationship, then we do single in the direction that there is a use case for. I do not think it hurts to make it bidirectional right away. It gives users of the class more options to choose from.

Reply:
@kulgan updated the PR to have relationships from `Study` to its children.  What's a good way to add the relationships from children to study?  Is that what `backref` does?  I'd also like the relationship from children to study to be lazy.

Reply:
see https://docs.sqlalchemy.org/en/13/orm/basic_relationships.html

Reply:
based on the in-person conversation, we agreed that adding relationships from study rule to its children is sufficient for our immediate use case.   if a future use case arises where a back reference is needed, we'll add it at that time.